core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -584,16 +590,20 @@ public void runInContext() {
 
     @Override
     public void closed(final Status status) {
+      // For cancellations, promptly inform any users of the context that their work should be
+      // aborted. Otherwise, we can wait until pending work is done.
+      if (!status.isOk()) {
+        // The callExecutor might be busy doing user work. To avoid waiting, use an executor that
+        // is not serializing.
+        cancelExecutor.execute(new ContextCloser(context, status.getCause()));
+      }
       callExecutor.execute(new ContextRunnable(context) {
         @Override
         public void runInContext() {
-          try {
-            getListener().closed(status);
-          } finally {
-            // Regardless of the status code we cancel the context so that listeners
-            // are aware that the call is done.
+          if (status.isOk()) {
             context.cancel(status.getCause());
```

Comment:
We want to call the listener before cancelling the context. Basically, put this back the way it was, but with the new condition.